### ISSUE TITLE
Scrolling on titlebar using high resolution mouse scroll wheel is broken

### ISSUE THREAD

**Issue by RX14** [swaywm/sway#7473]
- **Sway Version:**
  - `master`

Ever since the new high-resolution scroll events have been implemented, scrolling to change windows in a tabbed container has been broken. I find this a really useful feature, so I wanted to fix it. Here's the libinput events coming from my MX Master 2S as you scroll one detent:

```
 event8   POINTER_SCROLL_WHEEL    +16.815s	vert 7.50/60.0* horiz 0.00/0.0 (wheel)
 event8   POINTER_SCROLL_WHEEL    +16.822s	vert 1.88/15.0* horiz 0.00/0.0 (wheel)
 event8   POINTER_SCROLL_WHEEL    +16.830s	vert 1.88/15.0* horiz 0.00/0.0 (wheel)
 event8   POINTER_SCROLL_WHEEL    +16.838s	vert 1.88/15.0* horiz 0.00/0.0 (wheel)
 event8   POINTER_SCROLL_WHEEL    +16.846s	vert 1.88/15.0* horiz 0.00/0.0 (wheel)
```

These add up to the 120 units required for "one click" on a standard mouse, but they are delivered in multiple events  (you can read the heuristic [here](https://gitlab.freedesktop.org/libinput/libinput/-/merge_requests/700)). The existing code for handling tabbar scroll is [here](https://github.com/swaywm/sway/blob/9498e4d2614fce6b29b64fd8a3861c1fb5be4b87/sway/input/seatop_default.c#L755), but it doesn't accumulate separate events together so scrolling only works when an event with 60 or greater axis units arrives.

The mouse theoretically has an option to turn off high resolution scrolling, but I think ever since libinput improved it's support for this mouse, it forces high-resolution scrolling on, so there seems little option to workaround this other than to fix sway.

I made a shit patch to test a few strategies for dealing with this, and came up with a good heuristic, though obviously this isn't upstreamable as-is. I just wanted to put in an issue here first to discuss instead of surprise the dev team with a MR.

```diff
--- a/sway/input/seatop_default.c
+++ b/sway/input/seatop_default.c
@@ -751,8 +751,26 @@ static void handle_pointer_axis(struct sway_seat *seat,
 			struct sway_node *active =
 				seat_get_active_tiling_child(seat, tabcontainer);
 			list_t *siblings = container_get_siblings(cont);
+
+			static float tabbar_scroll = 0;
+			static struct timespec tabbar_last_event;
+
+			// Reset tabbar_scroll after 1s passes
+			struct timespec now;
+			clock_gettime(CLOCK_MONOTONIC, &now);
+			float diff = (now.tv_sec - tabbar_last_event.tv_sec) + (now.tv_nsec - tabbar_last_event.tv_nsec) / 1000000000.0 ;
+			if (diff >= 1.0) {
+				tabbar_scroll = 0;
+			}
+			tabbar_last_event = now;
+
+			tabbar_scroll += scroll_factor * event->delta_discrete;
+
 			int desired = list_find(siblings, active->sway_container) +
-				roundf(scroll_factor * event->delta_discrete / WLR_POINTER_AXIS_DISCRETE_STEP);
+				truncf(tabbar_scroll / WLR_POINTER_AXIS_DISCRETE_STEP);
+
+			tabbar_scroll = fmod(tabbar_scroll, WLR_POINTER_AXIS_DISCRETE_STEP);
+
 			if (desired < 0) {
 				desired = 0;
 			} else if (desired >= siblings->length) {
```

**Comment by rpigott** [comment]
Dupe of #7338?

**Comment by RX14** [comment]
No, this is about sway's own scroll landling, not how events are sent to applications.

**Comment by mehdisadeghi** [comment]
>  events coming from my MX Master 2S 

@RX14 fwiw I had the exact issue with my MX Anywhere 3. I was using a Unified Receiver. I switched to directly connecting the mouse with my system's bluetooth and it solved the issue for me.

**Comment by emmanueltouzery** [comment]
I'm also seeing a regression on the titlebar mouse wheel scrolling since i upgraded from sway 1.7-4.fc37 to 1.8.1-1.fc38 (the scrolling is not reliable, it's missing events).
I'm also using a USB unifying receiver, and the mouse is a MX Master. I'll try switching to bluetooth.

**Comment by emmanueltouzery** [comment]
unfortunately switching to plain bluetooth without USB receiver doesn't seem to change anything, I still have the issue.

**Comment by RX14** [comment]
Reminder that I am seeking feedback for my above patch in regards to whether there is a better algorithm anyone can imagine, and where I should put the state storage required for an upstreamable patch. Once i can get an assurance that many time spent producing a "real" patch won't be wasted I'll put in a MR.

**Comment by emersion** [comment]
`wlr_seat` has logic to accumulate high-res events. I wonder if it would make sense to extract as a helper that Sway could use internally too? Or whether we should provide something at the `wlr_pointer` level?

**Comment by khamer** [comment]
Similar experience here using Logitech M720 mouse.

How I'd describe it is that trying to scroll past more than one tab, it seems to 'stick' on the first tab unless I stop scrolling then scroll more. I wonder if there's some functionality that's just counting all scrolling as a single scroll. wev shows separate events, and by switching to Bluetooth from the unifying receiver it does seem to workaround the issue.

Also, waybar (switching workspaces) does not seem to have this problem, just when switching sway tabs. I can replicate the issue with all terminals, browsers, or a mix, so I don't think it has anything to do with the applications in the tabs.

**Comment by khamer** [comment]
Still seeing this issue on a clean install with 1.8.1. Happy to gather whatever information I can to figure out a fix or workaround.

**Comment by roshal** [comment]
the workaround for me is to disable `REL_WHEEL_HI_RES` event via libinput quirks

https://wayland.freedesktop.org/libinput/doc/latest/device-quirks.html

`/etc/libinput/local-overrides.quirks`

```properties
[Logitech G604 USB]
MatchUdevType=mouse
MatchBus=usb
MatchVendor=0x046D
MatchProduct=0x4085
AttrEventCode=-REL_WHEEL_HI_RES
```

```properties
[Logitech MX Master 3 USB]
MatchUdevType=mouse
MatchBus=usb
MatchVendor=0x046D
MatchProduct=0x4082
AttrEventCode=-REL_WHEEL_HI_RES
```

quirks for @RX14

```properties
[Logitech MX Master 2S USB]
MatchUdevType=mouse
MatchBus=usb
MatchVendor=0x046D
MatchProduct=0x4069
AttrEventCode=-REL_WHEEL_HI_RES
```

quirks for @mehdisadeghi

```properties
[Logitech MX Anywhere 3 USB]
MatchUdevType=mouse
MatchBus=usb
MatchVendor=0x046D
MatchProduct=0x4090
AttrEventCode=-REL_WHEEL_HI_RES
```

quirks for @khamer

```properties
[Logitech M720 USB]
MatchUdevType=mouse
MatchBus=usb
MatchVendor=0x046D
MatchProduct=0x405E
AttrEventCode=-REL_WHEEL_HI_RES
```

libinput quirks

https://gitlab.freedesktop.org/libinput/libinput/blob/main/quirks

libratbag devices

https://github.com/libratbag/libratbag/tree/master/data/devices

**Comment by RX14** [comment]
I'd rather just someone clean up the patch I already posted and get it upstream. I am currently far too busy but it's a simple patch, the remaining work is mostly finding out a place to store some state.

**Comment by roshal** [comment]
**wlroots** [`wlr_seat_pointer`](https://gitlab.freedesktop.org/wlroots/wlroots/blob/0.17.2/types/seat/wlr_seat_pointer.c) has [`update_value120_accumulators`](https://gitlab.freedesktop.org/wlroots/wlroots/blob/0.17.2/types/seat/wlr_seat_pointer.c#L289) to accumulate the axis value
